LeaderboardBackend.Test/Features/Users/RegistrationTests.cs

```diff
@@ -182,25 +171,126 @@ public async Task Register_UsernameAlreadyTaken_ReturnsConflictAndErrorCode()
+
         RegisterRequest createExistingUserReq = _registerReqFaker.Generate();
-        await Client.PostAsJsonAsync(Routes.REGISTER, createExistingUserReq);
+        await client.PostAsJsonAsync(Routes.REGISTER, createExistingUserReq);
+
         RegisterRequest request = _registerReqFaker.Generate() with { Email = createExistingUserReq.Email.ToLower() };
+        HttpResponseMessage res = await client.PostAsJsonAsync(Routes.REGISTER, request);
 
-        HttpResponseMessage res = await Client.PostAsJsonAsync(Routes.REGISTER, request);
+        res.Should().HaveStatusCode(HttpStatusCode.Accepted);
 
-        res.Should().HaveStatusCode(HttpStatusCode.Conflict);
-        ValidationProblemDetails? content = await res.Content.ReadFromJsonAsync<ValidationProblemDetails>();
-        content.Should().NotBeNull();
-        content!.Errors.Should().BeEquivalentTo(new Dictionary<string, string[]>
+        emailSender.Verify(s =>
+            s.EnqueueEmailAsync(
+                createExistingUserReq.Email,
+                "Confirm Your Account",
+                It.IsAny<string>()
+            ),
+            Times.Exactly(2)
+        );
+    }
+
+    [TestCase(UserRole.Confirmed)]
+    [TestCase(UserRole.Administrator)]
+    [TestCase(UserRole.Banned)]
+    public async Task Register_EmailAlreadyUsed_OtherRoles(UserRole role)
+    {
+        Mock<IEmailSender> emailSender = new();
+
+        using HttpClient client = _factory.WithWebHostBuilder(builder =>
+            builder.ConfigureTestServices(services =>
+                services.AddScoped(_ => emailSender.Object)
+        )).CreateClient();
+
+        IServiceScope scope = _factory.Services.CreateScope();
+        IUserService service = scope.ServiceProvider.GetRequiredService<IUserService>();
+        ApplicationContext context = scope.ServiceProvider.GetRequiredService<ApplicationContext>();
+
+        CreateUserResult result = await service.CreateUser(new()
         {
-            { nameof(RegisterRequest.Email), new[] { "EmailAlreadyUsed" } }
+            Email = $"testregister.emailused.{role}@example.com",
+            Password = "P4ssword",
+            Username = $"RegisterTestEmailUsed{role}",
         });
+
+        result.IsT0.Should().BeTrue();
+        User user = result.AsT0;
+        context.Update(user);
+        user!.Role = role;
+
+        await context.SaveChangesAsync();
+
+        RegisterRequest request = _registerReqFaker.Generate() with { Email = $"testregister.emailused.{role}@example.com" };
+        HttpResponseMessage res = await client.PostAsJsonAsync(Routes.REGISTER, request);
+
+        res.Should().HaveStatusCode(HttpStatusCode.Accepted);
+
+        emailSender.Verify(s =>
+            s.EnqueueEmailAsync(
+                $"testregister.emailused.{role}@example.com",
+                "A Registration Attempt was Made with Your Email",
+                It.IsAny<string>()
+            ),
+            role is UserRole.Banned ? Times.Never() : Times.Once()
```

Comment:
[Optional] Make the number of times a parameter for the test.

Reply:
```cs
[TestCase(UserRole.Confirmed, 1)]
[TestCase(UserRole.Administrator, 1)]
[TestCase(UserRole.Banned, 0)]
public async Task Register_EmailAlreadyUsed_OtherRoles(UserRole role, int times)
{
    // snip
    Times.Exactly(times)
```

Reply:
I was so laser-focused on the named methods like `Once()` I forgot about `Exactly()` 🤦‍♂️